src/languages/latex.js

```diff
@@ -1,70 +1,248 @@
-              variants: [
-                {begin: /\[/, end: /\]/},
-                {begin: /\{/, end: /\}/}
-              ]
-            },
-            {
-              begin: /\s*=\s*/, endsWithParent: true,
-              relevance: 0,
-              contains: [
-                {
-                  className: 'number',
-                  begin: /-?\d*\.?\d+(pt|pc|mm|cm|in|dd|cc|ex|em)?/
-                }
-              ]
-            }
-          ]
-        }
+        endsParent: true,
+        begin: KNOWN_CONTROL_WORDS
+      },
+      {
+        endsParent: true,
+        begin: L3_REGEX
+      },
+      {
+        endsParent: true,
+        variants: DOUBLE_CARET_VARIANTS
+      },
+      {
+        endsParent: true,
+        relevance: 0,
+        variants: L2_VARIANTS
       }
     ]
   };
-
-  return {
-    name: 'LaTeX',
-    aliases: ['tex'],
-    contains: [
-      COMMAND,
+  const MACRO_PARAM = {
+    className: 'params',
+    relevance: 0,
+    begin: /#+\d?/
+  };
+  const DOUBLE_CARET_CHAR = {
+    // relevance: 1
+    variants: DOUBLE_CARET_VARIANTS
+  };
+  const SPECIAL_CATCODE = {
+    className: 'built_in',
+    relevance: 0,
+    begin: /[$&^_]/
+  };
+  const MAGIC_COMMENT = {
+    className: 'meta',
+    begin: '% !TeX',
+    end: '$',
+    relevance: 10
+  };
+  const COMMENT = hljs.COMMENT(
+    '%',
+    '$',
+    {
+      relevance: 0
+    }
+  );
+  const EVERYTHING_BUT_VERBATIM = [
+    CONTROL_SEQUENCE,
+    MACRO_PARAM,
+    DOUBLE_CARET_CHAR,
+    SPECIAL_CATCODE,
+    MAGIC_COMMENT,
+    COMMENT
+  ];
+  const BRACE_GROUP_NO_VERBATIM = {
+    begin: /\{/, end: /\}/,
+    relevance: 0,
+    contains: ['self', ...EVERYTHING_BUT_VERBATIM]
+  };
+  const ARGUMENT_BRACES = hljs.inherit(
+    BRACE_GROUP_NO_VERBATIM,
+    {
+      relevance: 0,
+      endsParent: true,
+      contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_VERBATIM]
+    }
+  );
+  const ARGUMENT_BRACKETS = {
+    begin: /\[/,
+      end: /\]/,
+    endsParent: true,
+    relevance: 0,
+    contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_VERBATIM]
+  };
+  const SPACE_GOBBLER = {
+    begin: /\s+/,
+    relevance: 0
+  };
+  const ARGUMENT_M = [ARGUMENT_BRACES];
+  const ARGUMENT_O = [ARGUMENT_BRACKETS];
+  const ARGUMENT_AND_THEN = function(arg, starts_mode) {
+    return {
+      contains: [SPACE_GOBBLER],
+      starts: {
+        relevance: 0,
+        contains: arg,
+        starts: starts_mode
+      }
+    };
+  };
+  const CSNAME = function(csname, starts_mode) {
+    return {
+        begin: '\\\\' + csname + '(?![a-zA-Z@:_])',
+        keywords: {$pattern: /\\[a-zA-Z]+/, keyword: '\\' + csname},
+        relevance: 0,
+        contains: [SPACE_GOBBLER],
+        starts: starts_mode
+      };
+  };
+  const BEGIN_ENV = function(envname, starts_mode) {
+    return hljs.inherit(
       {
-        className: 'formula',
-        contains: [COMMAND],
+        begin: '\\\\begin(?=\\s*\\r?\\n?\\s*\\{' + envname + '\\})',
```

Comment:
We'll need to change this, this can result in a polynomial backtracking. Because the `\s*` will dual with each other since there is no guarantee there will be any content in the middle.

Can this be replaced with a single group? `(\s\r\n)` ?  If not we may need two different paths here to avoid backtracking.

Reply:
Yes, I think it can. The real rule is that there may only be one newline, but that's really not so important after `\begin`. I'll try to make a PR shortly.

(Sorry for being so unresponsive lately. It's been difficult to find some time to work on this beside my other obligations.. There will be different times again, though.)

Reply:
I took a quick pass:

```ts
begin: '\\\\begin(?=[ \t]*(\\r?\\n[ \t]*)?\\{' + envname + '\\})',
```

Thoughts?

Passes all your tests.

Reply:
Yes, that's what it should be. (I think I thought `\s` didn't capture newlines, which seems to also be wrong.) Don't know if there are any (dis)advantages to making the group non-capturing performance-wise..?

Reply:
> Don't know if there are any (dis)advantages to making the group non-capturing performance-wise..?

I've seen mixed results on this, so typically I don't worry about captures unless I'm using references or one of the few rules that uses reference.